lib/collections.c

```diff
@@ -231,3 +235,157 @@ void list_free(List* list)
+}
+
+static void dictionary_rebuild(Dictionary* dictionary)
+{
+    size_t new_max_length;
+    if (dictionary->length == dictionary->max_length)
+    {
+        new_max_length = dictionary->max_length * 2;
+    }
+    else if (dictionary->length * 4 <= dictionary->max_length &&
+             dictionary->max_length > INITIAL_DICTIONARY_MAX_LENGTH)
+    {
+        new_max_length = dictionary->max_length / 2;
+    }
+    else
+    {
+        return;
+    }
+
+    KeyValuePair* new_key_value_pairs =
+        calloc(new_max_length, sizeof(KeyValuePair));
+    if (new_key_value_pairs == NULL)
+    {
+        fprintf(stderr, "Error: dictionary_rebuild: calloc failed\n");
+        exit(EXIT_FAILURE);
+    }
+    for (size_t i = 0; i < dictionary->max_length; i++)
+    {
+        KeyValuePair key_value_pair = dictionary->key_value_pairs[i];
+        uint64_t index = hash_key(key_value_pair.key) % new_max_length;
+        while (new_key_value_pairs[index].key != NULL)
+        {
+            index = (index + 1) % dictionary->max_length;
+        }
+        new_key_value_pairs[index] = key_value_pair;
+    }
+    dictionary->max_length = new_max_length;
+    dictionary->key_value_pairs = new_key_value_pairs;
+}
+
+void dictionary_insert(Dictionary* dictionary, const char* key, void* value)
+{
+    KeyValuePair key_value_pair = {key, value};
+    dictionary_rebuild(dictionary);
+    uint64_t index = hash_key(key) % dictionary->max_length;
+    while (dictionary->key_value_pairs[index].key != NULL)
+    {
+        index = (index + 1) % dictionary->max_length;
+    }
+    dictionary->key_value_pairs[index] = key_value_pair;
+    dictionary->length++;
+}
+
+void* dictionary_get(Dictionary* dictionary, const char* key)
+{
+    uint64_t index = hash_key(key) % dictionary->max_length;
+    while (dictionary->key_value_pairs[index].key != NULL)
+    {
+        if (strcmp(dictionary->key_value_pairs[index].key, key) == 0)
+        {
+            return dictionary->key_value_pairs[index].value;
+        }
+        index = (index + 1) % dictionary->max_length;
+    }
+    fprintf(stderr, "Error: dictionary_get: key not found\n");
+    exit(EXIT_FAILURE);
+}
+
+void* dictionary_remove(Dictionary* dictionary, const char* key)
+{
+    uint64_t index = hash_key(key) % dictionary->max_length;
+    while (dictionary->key_value_pairs[index].key != NULL)
+    {
+        if (strcmp(dictionary->key_value_pairs[index].key, key) == 0)
+        {
+            void* value = dictionary->key_value_pairs[index].value;
+            dictionary->key_value_pairs[index].key = NULL;
+            dictionary->key_value_pairs[index].value = NULL;
+            dictionary->length--;
+            dictionary_rebuild(dictionary);
+            return value;
+        }
+        index = (index + 1) % dictionary->max_length;
+    }
+    fprintf(stderr, "Error: dictionary_remove: key not found\n");
+    exit(EXIT_FAILURE);
+}
+
+bool dictionary_contains(Dictionary* dictionary, const char* key)
```

Comment:
Figured this might be useful for the caller of the dictionary interface functions since we're throwing errors when keys don't exist (example usage: `if (dictionary_contains(...)) { dictionary_remove(...) }`)